### PR TITLE
ENTESB-12645: Extend db backup to cover external databases

### PR DESCRIPTION
**Pull request by phantomjinx**
* cmd/.../backup.go
 * operator-config switch required for command line execution
 * requires installed syndesis CR to be passed in

* install_forge.go
 * Tidy up with centralising of code into util functions

* cmd/.../restore.go
 * operator-config switch required for command line execution
 * custom-options switch made available for command line execution if
   alternative custom options to pg_restore are preferred
 * requires installed syndesis CR to be passed in

* backup/syndesis...yml.tmpl
 * Templates for Jobs to be executed by backup and restore

* backup/[backup|restore].go
 * Major refactoring to generate a Job which creates a delegate Pod for
   running pg_dump/pg_restore which both connect remotely to the db

* configuration.go
 * BackupConfiguration provides conduit between code and templates

### CODE REVIEW

**Comment by heiko-braun** [comment]
@phantomjinx Looks like this needs to be rebased.

File: install/operator/pkg/syndesis/backup/backup.go [hunk 1]
```diff
@@ -34,30 +35,54 @@ import (
...
+	backupPath      string
+	Namespace       string
+	BackupDir       string
+	Delete          bool
+	LocalOnly       bool
+	Context         context.Context
+	Client          *client.Client
+	Syndesis        *v1beta1.Syndesis
+	CustomOptions   string
+	synConfig       *configuration.Config
```
Comment (lgarciaaco) [inline install/operator/pkg/syndesis/backup/backup.go]: I believe `sysConfig` doesn't need to use pointer semantic?

File: install/operator/pkg/syndesis/backup/backup.go [hunk 2]
```diff
@@ -193,35 +218,273 @@ func (b *Backup) ensureDir() (err error) {
...
+// Execute a task within a Job
+//
+func (b *Backup) executeJob(jobName string, jobTask jobTaskFunc) error {
+	//
+	// TODO
+	//
+	// Since the job keeps dying after 8-12 seconds regardless of the pod
+	// still running (maybe a 3.11 issue?), it is necessary to monitor the
+	// target pod instead. Is the job is more trouble than its worth??
+	//
```
Comment (lgarciaaco) [inline install/operator/pkg/syndesis/backup/backup.go]: Is this really the case? Very fishy, did you try 4.x?

File: install/operator/pkg/util/kube.go [hunk 3]
```diff
@@ -173,3 +180,33 @@ func Exec(o ExecOptions) error {
...
+			CONTROLLER_UID_LABEL: controllerUid,
+			JOB_NAME_LABEL:       job.Name,
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return podList, nil
+}
```
Comment (lgarciaaco) [inline install/operator/pkg/util/kube.go]: The method is called only once from `backup,go`, no reason for having it here instead of in the package where it is called from, actually it is totally fine not to have a method for this since it is called once.

File: install/operator/pkg/util/kube.go [hunk 4]
```diff
@@ -173,3 +180,33 @@ func Exec(o ExecOptions) error {
...
+
+/**
+ * Refreshes a pod with its latest state on the server
+ */
+func RefreshPod(ctx context.Context, c client.Client, pod *corev1.Pod) error {
```
Comment (lgarciaaco) [inline install/operator/pkg/util/kube.go]: Similar comment to previous one

File: install/operator/pkg/syndesis/backup/backup.go [hunk 5]
```diff
@@ -193,35 +218,273 @@ func (b *Backup) ensureDir() (err error) {
...
+			//
+			return true, nil
+		}
+
+		//
+		// Still waiting on pods being created by job
+		//
+		return false, nil
+	})
+
```
Comment (lgarciaaco) [inline install/operator/pkg/syndesis/backup/backup.go]: The wait into a wait approach is confusing. Can't we really use the job to figure out these stuff? If we can't use the job for this case, then it is probably better to spin up a pod directly because using a job that 'keeps dying' turns from being handy to being a trouble.

File: install/operator/pkg/syndesis/backup/restore.go [hunk 6]
```diff
@@ -52,6 +65,10 @@ func (b *Backup) Restore() (err error) {
 			return err
 		}
 
+		nfi, err := os.Stat(dir)
+		if !nfi.IsDir() {
+			return fmt.Errorf("Unzipped backup directory does not exist: %s", dir)
+		}
 		b.BackupDir = dir
```
Comment (lgarciaaco) [inline install/operator/pkg/syndesis/backup/restore.go]: Why is this needed, dir is surely created few lines above or?
```
		dir, err := ioutil.TempDir("/tmp", "restore-")
		if err != nil {
			return err
		}
		defer os.RemoveAll(dir)
		os.Chmod(dir, 0755)
```

File: install/operator/pkg/syndesis/backup/restore.go [hunk 7]
```diff
@@ -63,9 +80,9 @@ func (b *Backup) Restore() (err error) {
 		return
 	}
 
-	if err = b.RestoreResources(); err != nil {
-		return
-	}
+	// if err = b.RestoreResources(); err != nil {
+	// 	return
+	// }
```
Comment (lgarciaaco) [inline install/operator/pkg/syndesis/backup/restore.go]: Did you forget to uncomment this?

File: install/operator/pkg/syndesis/backup/backup.go [hunk 8]
```diff
@@ -290,15 +548,15 @@ func (b *Backup) clientConfig() *rest.Config {
...
 		cl, err := client.New(b.clientConfig(), client.Options{})
 		if err != nil {
-			return nil, err
+			util.ExitOnError(err)
 		}
 		b.Client = &cl
 	}
-	return *b.Client, nil
+	return *b.Client
 }
```
Comment (lgarciaaco) [inline install/operator/pkg/syndesis/backup/backup.go]: I don't think this is a good approach and this is why. If the operator is scheduled to run a backup(and we can afford that a backup fails, because it is a backup right) and this method fails for whatever reason, then this will kill the operator.

Rather wrap the error and return I would say.

File: install/operator/pkg/util/syndesis_object.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
+/*
```
Comment (lgarciaaco) [inline install/operator/pkg/util/syndesis_object.go]: I don't think this file should exist at all

File: install/operator/pkg/syndesis/backup/restore.go (hunk 6 above)
Comment (phantomjinx) [reply install/operator/pkg/syndesis/backup/restore.go]: Existing code for handling a backup directory that has been zipped up. On the command line it is possible to restore by specifying an aforementioned zip archive.

**Comment by phantomjinx** [comment]
Hey @zregvart @heiko-braun 
In Luis' absence, would you mind giving this update a once-over, please?

As far as tests go, many more could be added but I really need to talk to Luis first in order to understand how to approach unit testing functions such as Backup.Run(). Of course, if you have any input on that I'd be really interested.

Thanks

PGR
